.github/workflows/build-and-test.yml

```diff
@@ -81,53 +44,26 @@ jobs:
           # We need to fetch all branches and commits so that Nx affected has a base to compare against.
           fetch-depth: 0
 
-      - uses: actions/checkout@v3
+      - uses: actions/checkout@v4
         name: Checkout [Default Branch]
         if: ${{ github.event_name != 'pull_request' }}
         with:
           # We need to fetch all branches and commits so that Nx affected has a base to compare against.
           fetch-depth: 0
 
-      - name: Set up dependencies
-        uses: ./.github/actions/setup
+      - name: Derive appropriate SHAs for base and head for `nx affected` commands
+        uses: nrwl/nx-set-shas@v4
 
       - name: Initialize the Nx Cloud distributed CI run
-        run: yarn nx-cloud start-ci-run
+        run: npx nx-cloud start-ci-run --distribute-on="./nx/workflows/dynamic-changesets.yml" --stop-agents-after="e2e"
+
+      - name: Set up dependencies
+        uses: ./.github/actions/setup
 
       - name: Run commands in parallel
         run: |
-          pids=()
-          # list of commands to be run on main has env flag NX_CLOUD_DISTRIBUTED_EXECUTION set to false
-          NX_CLOUD_DISTRIBUTED_EXECUTION=false yarn nx-cloud record -- yarn nx format:check & pids+=($!)
-
-          # list of commands to be run on agents
-          yarn nx affected -t lint --parallel=3 &
-          pids+=($!)
-
-          yarn nx affected -t test --parallel=3 &
-          pids+=($!)
-
-          yarn nx affected -t build --exclude=docs --parallel=3 &
-          pids+=($!)
-
-          yarn nx affected -t e2e --parallel=1 &
-          pids+=($!)
-
-          yarn nx affected -t component-test --parallel=1 &
-          pids+=($!)
-
-          # run all commands in parallel and bail if one of them fails
-          for pid in ${pids[*]}; do
-            if ! wait $pid; then
-              exit 1
-            fi
-          done
-
-          exit 0
-
-      - name: Stop Nx Cloud DTE agents
-        if: ${{ always() }}
-        run: yarn nx-cloud stop-all-agents
+          npx nx-cloud record -- npx nx format:check
+          npx nx affected -t lint build test component-test e2e --parallel=4 --exclude=docs
```

Comment:
suggestion: Remove the `--parallel` option as distribution is optimized by Nx Agents based on the activated changeset.

```suggestion
          npx nx affected -t lint build test component-test e2e --exclude=docs
```

Reply:
thanks!

Reply:
@LayZeeDK I don't see any evidence about `--parallel` is a bad idea for cloud agents. In fact, tried it on my private projects.

![image](https://github.com/user-attachments/assets/c06b562c-48e7-4b0c-af61-b6a68bd36d67)

You can see that the agent is running 4 tasks in parallel, as I am running it with `--parallel=4`. Imo it should be better to have them run things in parallel, wdyt?

Reply:
Are you using Nx Agents in your private project?

```shell
nx-cloud start-ci-run --distribute-on="./nx/workflows/dynamic-changesets.yml"
```

Reply:
@LayZeeDK yes!